fuzz/fuzz_targets/native_c.rs

```diff
@@ -0,0 +1,140 @@
+#![no_main]
+
+use std::ffi::{CStr, CString};
+use std::fmt::Write;
+
+use libfuzzer_sys::fuzz_target;
+
+#[derive(Debug)]
+enum State<'a> {
+    Ok(&'a str),
+    Overflow,
+}
+
+pub struct FormatBuf<'a> {
+    buf: &'a mut String,
+    max_len: usize,
+}
+
+impl<'a> Write for FormatBuf<'a> {
+    fn write_str(&mut self, s: &str) -> Result<(), std::fmt::Error> {
+        if self.buf.len() + s.len() > self.max_len {
+            return Err(std::fmt::Error);
+        }
+        self.buf.push_str(s);
+        Ok(())
+    }
+}
+
+fn asciify(x: &str) -> String {
+    let mut result = String::with_capacity(x.len() * 4);
+    for ch in x.chars() {
+        if ch.is_ascii() {
+            result.push(ch);
+        } else {
+            write!(&mut result, "\\u{{{:x}}}", ch as u32).ok();
+        }
+    }
+    result
+}
+
+fuzz_target!(|data: &[u8]| {
+    if data.len() == 0 {
+        return;
+    }
+    let alternate = data[0] % 2 == 0;
+    let data = &data[1..];
+    let mut str_buf = String::with_capacity(16384);
+    let mut buf = [0u8; 4096];
+    let mut demangle = rustc_demangle_native_c::CDemangle::zero();
+    let state;
+    if let Ok(s) = std::str::from_utf8(data) {
```

Comment:
Should we require utf-8? Or maybe we should have a separate fuzz target which just throws arbitrary input at the C impl without verifying outputs or so (or unit tests with assertions of errors on non-ascii input?)

Reply:
the C code requires ascii-only pretty early